var/spack/repos/builtin/packages/openpmd-api/package.py

```diff
@@ -0,0 +1,99 @@
+##############################################################################
+# Copyright (c) 2013-2018, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+
+class OpenpmdApi(CMakePackage):
+    """API for easy reading and writing of openPMD files"""
+
+    homepage = "http://www.openPMD.org"
+    url      = "https://github.com/openPMD/openPMD-api/archive/1.0.0.tar.gz"
+    maintainers = ['ax3l']
+
+    version('develop', branch='dev',
+            git='https://github.com/openPMD/openPMD-api.git')
+
+    variant('mpi', default=True,
+            description='Enable parallel I/O')
+    variant('hdf5', default=True,
+            description='Enable HDF5 support')
+    variant('adios1', default=False,
+            description='Enable ADIOS1 support')
+    variant('adios2', default=False,
+            description='Enable ADIOS2 support')
+    # variant('json', default=True,
+    #         description='Enable JSON support')
+    variant('python', default=True,
+            description='Enable Python bindings')
+    variant('test', default=True,
+            description='Build the tests')
+
+    depends_on('cmake@3.10.0:', type='build')
+    depends_on('boost@1.62.0:')
+    depends_on('mpark-variant@1.3.0:')
+    depends_on('catch@2.2.1: ~single_header', when='+test', type='build')
+    depends_on('mpi@2.3:', when='+mpi')  # might become MPI 3.0+
+    depends_on('hdf5@1.8.6: ~mpi', when='~mpi +hdf5')
+    depends_on('hdf5@1.8.6: +mpi', when='+mpi +hdf5')
+    depends_on('adios@1.10.0: ~mpi', when='~mpi +adios1')
+    depends_on('adios@1.10.0: +mpi', when='+mpi +adios1')
+    depends_on('adios2@2.1.0: ~mpi', when='~mpi +adios2')
+    depends_on('adios2@2.1.0: +mpi', when='+mpi +adios2')
```

Comment:
Dependency forwarding is a controversial subject. The general consensus has been not to forward dependencies and let the user handle it themselves. The concretizer doesn't handle this well at the moment, leading to a lot of bugs.

Reply:
yep I understand that. the idea here is as follows: a user wants to use this I/O library either with or without MPI *at all*. If any of the deps pulls in MPI, e.g. python startups and program startups will trigger MPI deps which is very confusing to users.

Since our default is `+mpi` for all variants, a newcomer that wants to use the simple python bindings of the lib with vanilla spack would otherwise have to type:
```
# the other variants will be default True in the future as well as soon as implemented
spack install openPMD-api ~mpi ^hdf5~mpi ^adios1~mpi ^adios2~mpi
spack load openPMD-api ~mpi ^hdf5~mpi ^adios1~mpi ^adios2~mpi
```

or manually edit the `packages.yaml` variant defaults for all dependencies (no CLI yet) - all in all too cumbersome to start when people are used to one-liners.

With a simple
```
spack install openPMD-api ~mpi
spack load openPMD-api ~mpi
```
and
```
spack install openPMD-api
spack load openPMD-api
```

this is something a user can understand, so I really would like to use forwarding here.

Reply:
Ouch yes, thank you so much.

Oh dang, i forgot that `depends_on('package +variant')` is currently broken.
Yes you are right - even if I wanted to forward variants (and I want to) it will not work...

Thanks for the great deails @adamjstewart, I will update the PR for now.